Improve unicode handling in Python 2

darkrain42 opened the pull request:
Improve pyldap's handling of `unicode` strings containing non-ASCII characters (namely #54 and #39) that result from Python 2's "s" format unit trying to coerce `unicode` objects to a `char*` by way of the `ascii` encoding.

After looking at @koniiiik's diagnosis of the issues and suggestions in #54 (which was an immense time-saver), I decided that letting Python do most of the heavy lifting and memory management was best approach -- using the "es" format unit in the C bindings would have introduced the need for explicit memory management, and increased the ongoing maintenance burden (more changes across a number of the C bindings).

With the changes here:

- In Python 2, "text" values passed from Python to the C bindings are expected to be UTF-8 encoded `bytes` objects
- In Python 3, "text" values passed from Python to the C bindings are expected to be `unicode` objects, and we rely on Python to coerce that into a utf-8 representation (and handle memory management).

I also fixed two other issues while I was doing this, namely a memory leak in `attrs_from_List` (Python object leak) and the `subschemasubentry` DN decoding not actually working the way it's commented under Python 3 (`_unbytesify_value` was explicitly a no-op in Python 3, so the value was returned as a bytes object, rather than a string).

darkrain42 committed:
tox.ini: Add newer Python versions

darkrain42 committed:
Add (failing) tests for Unicode bind / DN encoding

References #39, #54

darkrain42 committed:
Decode subschemasubentry DN as described in Python 3

_unbytesify_value is a no-op in Python 3, so add a new function that actually
decodes bytes to text when bytes_mode is set to False, reagrdless of Python
version.

To match the other bytesify functions, the type assertion is only run under
Python 2.

darkrain42 committed:
PY2: Explicitly encode to UTF-8 before calling C

Rework how pyldap handles converting Unicode to UTF-8 encoded byte-strings to
avoid inconsistencies between Python 2 and Python 3 in how Python's C API
handles the "s" format unit. The inconsistencies result in Python 2 raising
UnicodeEncodeErrors when Unicode strings with non-ASCII characters are passed
to the C bindings.

- In Python 3, the "s" format expects the argument to be a unicode string, and
  it encodes it using UTF-8 (unconditionally).
- In Python 2, the "s" format tries to encode unicode strings using the
  default encoding (sys.getdefaultencoding(), which defaults to 'ascii').

Avoid these inconsistencies (while preserving the warnings/error handling) by
explicitly encoding values to UTF-8 bytes, instead of passing passing unicode
objects to C. Doing this addresses the issue, while still relying on Python
for GC/memory management. The alternatives (e.g. using the "es" format unit)
would require more invasive changes in the C bindings, including explicit
memory management.

Also update `attrs_from_List` to align with these changes:

- In Python 2, assume a sequence of bytes objects (encoded to UTF-8)
- In Python 3, assume a sequence of string (Unicode) objects and rely on
  Python to encode to UTF-8 and handle the memory allocations.

This simplification of attrs_from_List also fixes a leak of the bytes objects;
previously, the new object returned by PyUnicode_AsUTF8String was leaked.

Fixes #4
Fixes #39
Fixes #54

darkrain42 committed:
Add (failing) tests for search_subschemasubentry encoding

rbarrois commented:
Thanks for the contribution!

This looks quite interesting, however I'll need some time to read it all — especially with the inverted logic around ``_unbytesify_value`` / ``_bytesify_result``.

koniiiik commented:
I didn't make a very thorough review, but at a glance it looks good to me, and I can confirm that it does fix our issue with non-ASCII binds, and credentials. I really can't speak to some of the changes in the C code, though.

darkrain42 commented:
I don't know how much it helps, but in the final commit, the only functional change in `ldapobject.py` should be to `_unbytesify_value`; the rest is a search/replace renaming of that set of helper functions to reflect the changed behavior.  (I noticed the `subschemasubentry` issue while reviewing the search/replace diff, and went back and fixed that first.)

stardust85 commented:
Please merge soon. I'm affected by this bug.

rbarrois commented:
@stardust85 we'll do our best; but I'm not paid to work on this ­— as most open source projects, we work on a best-effort basis.

However, all contributions are welcome: if you've been able to test this code, or can review the changes, that helps us a lot!

Thanks for your understanding :)

darkrain42 committed:
Improve comments and make formatting consistent

PEP8-styling accidentally snuck into parts of _unbytesify_result.  Also
comment a commented-out bind test.

rbarrois reviewed (approved):
Thanks, this is awesome and pretty thorough.

I'll just add a note in our docstrings to clarify your logic :)
